pkg/agent/openflow/packetin.go

```diff
@@ -29,14 +29,20 @@ type PacketInHandler interface {
 const (
 	// Action explicitly output to controller.
 	ofprAction ofpPacketInReason = 1
+	// Action output to cluster network policy controller.
+	ofcnpAction ofpPacketInReason = 2
+	// Name for traceflow PacketInHandler
+	tfPacketHandlerName = "traceflow"
+	// Name for CNP Logging PacketInHandler
+	npPacketHandlerName = "networkpolicy"
 	// Max packetInQueue size.
 	packetInQueueSize int = 256
 )
 
 func (c *client) RegisterPacketInHandler(packetHandlerName string, packetInHandler interface{}) {
 	handler, ok := packetInHandler.(PacketInHandler)
 	if !ok {
-		klog.Errorf("Invalid Traceflow controller.")
+		klog.Errorf("Invalid controller.")
 		return
 	}
 	c.packetInHandlers[packetHandlerName] = handler
```

Comment:
Hi Qiyue,
I am looking at an issue that will use a similar mechanism: https://github.com/vmware-tanzu/antrea/issues/1004

In light of that and in general, I feel it is better to use reason in the map to support multiple handlers for one reason rather than just supporting one handler. Could map[reason]map[string/handlername]handler better?

Reply:
This sounds like a necessary change. What's the targeting reason or handler that issue is looking at?

Reply:
For a given "reason", there can be multiple handlers that want to process the punted packets. Handlers can be independent of each other. In the above case with CNP, logging will be one handler, and tracking denied/rejected connections can be another handler. We can implement them in a single handler but multiple handlers help in distinguishing functionality and a better generic design.

Reply:
Got it, that makes sense. Changing the map structure.